File: Published/SourceCode/Python/DoubleLinkedList.py

from API import List

class DoubleLinkedList(List):
    def __init__(self):
        self._head = None    # Pointer to list header
        self._tail = None    # Pointer to list tail
        self._listSize = 0   # Size of list

    def get(self, i):
        if not (0 <= i < self._listSize): raise IndexError("list index out of range")
        current = self._head
        for _ in range(i):
            current = current.next
        return current.elem

    def set(self, i, x):
        if not (0 <= i < self._listSize): raise IndexError("list index out of range")
        current = self._head
        for _ in range(i):
            current = current.next
        old = current.elem
        current.elem = x
        return old

    def add(self, i, x):
        if not (0 <= i <= self._listSize): raise IndexError("list index out of range")
        if self._listSize == 0:
            self._head = self._tail = DoubleLinkedListNode(x, None, None)
        elif i == 0:
            newhead = DoubleLinkedListNode(x, None, self._head)
            self._head.prev = newhead
            self._head = newhead
        elif i == self._listSize:
            newtail = DoubleLinkedListNode(x, self._tail, None)
            self._tail.next = newtail
            self._tail = newtail
        else:
            prev = self._head
            for _ in range(i-1):
                prev = prev.next
            next  = prev.next
            newnode = DoubleLinkedListNode(x, prev, next)
            prev.next = newnode
            next.prev = newnode
        self._listSize += 1

    def remove(self, i):
        if not (0 <= i < self._listSize): raise IndexError("list index out of range")
        if i == 0:
            removed = self._head
            self._head = removed.next
            self._head.prev = None
        elif i == self._listSize-1:
            removed = self._tail
            self._tail = removed.prev
            self._tail.next = None
        else:
            prev = self._head
            for _ in range(i-1):
                prev = prev.next
            removed = prev.next
            prev.next = removed.next
            prev.next.prev = prev
        removed.prev = removed.next = None   # For garbage collection
        self._listSize -= 1
        return removed.elem

    def isEmpty(self):
        return self._listSize == 0

    def size(self):
        return self._listSize

    def __iter__(self):
        current = self._head
        while current is not None:
            yield current.elem
            current = current.next


# Python does not have internal classes, so we have to make the list node standalone.
class DoubleLinkedListNode:
    def __init__(self, elem, prev, next):
        self.elem = elem   # Value for this node
        self.prev = prev   # Pointer to previous node in list
        self.next = next   # Pointer to next node in list


#######################################################################################
## What comes below is purely for debugging and testing purposes - it can be removed ##

def _printList(l):
    print("[", " ".join(str(e) for e in l), "]", l.size())

def _testList(l):
    if l.size() == 0:
        assert l._head is None
        assert l._tail is None
        return
    if l.size() == 1:
        assert l._head is l._tail
    assert l._head is not None
    assert l._tail is not None
    assert l._head.prev is None
    assert l._tail.next is None
    n = l._head
    while n is not None:
        if n.prev is not None:
            assert n.prev.next is n, (n.elem, n.prev, n.next)
        if n.next is not None:
            assert n.next.prev is n, (n.elem, n, n.next.prev)
        n = n.next

if __name__ == '__main__':
    a = DoubleLinkedList()
    _testList(a)
    print("Adding")
    for i in range(20):
        a.add(a.size(), chr(i+65))
        _testList(a)
        if i % 5 == 0:
            _printList(a)
    _printList(a)
    print("Setting")
    for i in range(0, a.size(), 2): a.set(i, a.get(i).lower())
    _printList(a)
    print("Removing")
    for _ in range(4):
        for i in range(a.size()-1, -1, -3):
            a.remove(i)
            _testList(a)
        _printList(a)

File: Published/SourceCode/Java/API/DoubleLinkedList.java

import java.util.Iterator;

class DoubleLinkedList<E> implements List<E> {
    private DNode head;     // Pointer to list header
    private DNode tail;     // Pointer to list tail
    private int listSize;   // Size of list

    DoubleLinkedList() {
        head = null;
        tail = null;
        listSize = 0;
    }

    private class DNode {
        E elem;       // Value for this node
        DNode prev;   // Pointer to previous node in list
        DNode next;   // Pointer to next node in list

        DNode(E elem, DNode prev, DNode next) {
            this.elem = elem;
            this.prev = prev;
            this.next = next;
        }
    }

    public E get(int i) {
        if (!(0 <= i && i < listSize)) throw new IndexOutOfBoundsException("list index out of range");
        DNode current = head;
        for (int k = 0; k < i; k++)
            current = current.next;
        return current.elem;
    }

    public E set(int i, E x) {
        if (!(0 <= i && i < listSize)) throw new IndexOutOfBoundsException("list index out of range");
        DNode current = head;
        for (int k = 0; k < i; k++)
            current = current.next;
        E old = current.elem;
        current.elem = x;
        return old;
    }

    public void add(int i, E x) {
        if (!(0 <= i && i <= listSize)) throw new IndexOutOfBoundsException("list index out of range");
        if (listSize == 0) {
            head = tail = new DNode(x, null, null);
        } else if (i == 0) {
            DNode newhead = new DNode(x, null, head);
            head.prev = newhead;
            head = newhead;
        } else if (i == listSize) {
            DNode newtail = new DNode(x, tail, null);
            tail.next = newtail;
            tail = newtail;
        } else {
            DNode prev = head;
            for (int k = 0; k < i-1; k++)
                prev = prev.next;
            DNode next = prev.next;
            DNode newnode = new DNode(x, prev, next);
            prev.next = newnode;
            next.prev = newnode;
        }
        listSize++;
    }

    public E remove(int i) {
        if (!(0 <= i && i < listSize)) throw new IndexOutOfBoundsException("list index out of range");
        DNode removed;
        if (i == 0) {
            removed = head;
            head = removed.next;
            head.prev = null;
        } else if (i == listSize-1) {
            removed = tail;
            tail = removed.prev;
            tail.next = null;
        } else {
            DNode prev = head;
            for (int k = 0; k < i-1; k++)
                prev = prev.next;
            removed = prev.next;
            prev.next = removed.next;
            prev.next.prev = prev;
        }
        removed.prev = removed.next = null;   // For garbage collection
        listSize--;
        return removed.elem;
    }

    public boolean isEmpty() {
        return listSize == 0;
    }

    public int size() {
        return listSize;
    }


    public Iterator<E> iterator() {
        return new DoubleLinkedListIterator();
    }

    private class DoubleLinkedListIterator implements Iterator<E> {
        private DNode current;
        DoubleLinkedListIterator() {
            current = head;
        }
        public boolean hasNext() {
            return current != null;
        }
        public E next() {
            E x = current.elem;
            current = current.next;
            return x;
        }
    }


/***************************************************************************************/
/** What comes below is purely for debugging and testing purposes - it can be removed **/

    public void _printList() {
        System.out.print("[ ");
        for (E e : this) System.out.print(e + " ");
        System.out.println("] " + size());
    }

    public void _testList() {
        if (size() == 0) {
            if (head != null) System.err.println("head is not null (size==0)");
            if (tail != null) System.err.println("tail is not null (size==0)");
            return;
        }
        if (size() == 1) {
            if (head != tail) System.err.println("head != tail (size==1)");
        }
        if (head == null) System.err.println("head is null");
        if (tail == null) System.err.println("tail is null");
        if (head.prev != null) System.err.println("head.prev is not null");
        if (tail.next != null) System.err.println("tail.next is not null");
        DNode n = head;
        while (n != null) {
            if (n.prev != null) {
                if (n.prev.next != n) System.err.println("n.prev.next != n");
            }
            if (n.next != null) {
                if (n.next.prev != n) System.err.println("n.next.prev != n");
            }
            n = n.next;
        }
    }

    public static void main(String[] args) {
        DoubleLinkedList<String> list = new DoubleLinkedList<>();
        list._testList();
        System.out.println("Adding");
        for (int i=0; i<20; i++) {
            list.add(list.size(), String.valueOf((char)(i+65)));
            list._testList();
            if (i % 5 == 0) list._printList();
        }
        list._printList();
        System.out.println("Setting");
        for (int i=0; i<list.size(); i+=2) list.set(i, list.get(i).toLowerCase());
        list._printList();
        System.out.println("Removing");
        for (int k=0; k<4; k++) {
            for (int i=list.size()-1; i>=0; i-=3) {
                list.remove(i);
                list._testList();
            }
            list._printList();
        }
    }
}
